Epmlstrtai 354

stelmakhmaksim opened the pull request:
Created main integration test that compares packages with classes generated manually and with classes generated by the page object generator.

AfanasievYuriy committed:
created module with basic logic for integrational tests

stelmakhmaksim committed:
Create basic integration test.

stelmakhmaksim committed:
Merge remote-tracking branch 'remotes/origin/EPMLSTRTAI-354' into EPMLSTRTAI-354

# Conflicts:
#	src/test/java/com/epam/page/object/generator/integrationalTests/Test1.java
#	src/test/resources/manual/page/DropdownMaterialize.java

stelmakhmaksim committed:
Merge branch 'master' of https://github.com/TAI-EPAM/page-objects-generator into EPMLSTRTAI-354

stelmakhmaksim committed:
Refactoring code.

stelmakhmaksim committed:
Refactoring code.

AfanasievYuriy committed:
created module with basic logic for integrational tests

stelmakhmaksim committed:
Main logic of testing is done.

KelinNik committed:
Add test methods with @Parameters

AfanasievYuriy committed:
Merge remote-tracking branch 'origin/EPMLSTRTAI-354' into EPMLSTRTAI-354

# Conflicts:
#	src/test/java/com/epam/page/object/generator/integrationalTests/Test1.java

AfanasievYuriy committed:
refactoring, parametrization added

AfanasievYuriy committed:
refactoring, basic logic confirmed; now testing class (MainIntegrationTest) can be shown to Alexander (for discussing purposes)

AfanasievYuriy committed:
refactoring, basic logic confirmed; now testing class (MainIntegrationTest) can be shown to Alexander (for discussing purposes)

stelmakhmaksim committed:
Ready to show to Alexander.

stelmakhmaksim committed:
Have a some problem with classloader.

AfanasievYuriy committed:
Changed a way to compile java files

AfanasievYuriy committed:
Added our own classloader;
Tests can be formed automaticly now (only need to add url to properties file);
Finally got working version (maybe need some refactoring)

stelmakhmaksim committed:
Merge branch 'EPMLSTRTAI-354' of D:\Programming\Java\Projects\page-objects-generator with conflicts.

AfanasievYuriy committed:
Added our own classloader;
Tests can be formed automaticly now (only need to add url to properties file);
Finally got working version (maybe need some refactoring)

stelmakhmaksim committed:
Merge remote-tracking branch 'origin/EPMLSTRTAI-354' into EPMLSTRTAI-354

AfanasievYuriy committed:
Edited regexp (for universal OS file path separator). now working on win and unix

stelmakhmaksim committed:
Merge remote-tracking branch 'origin/EPMLSTRTAI-354' into EPMLSTRTAI-354

stelmakhmaksim committed:
Global refactoring is done

hdghg commented on pom.xml:
```diff
@@ -89,6 +89,11 @@
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>RELEASE</version>
```
junit is usually added to test scope of project. Either add test scope or remove this dependency configuration at all (junit is in transitive dependencies of jdi-uitest-web)

AfanasievYuriy replied:
fixed (junit no more in pom.xml)

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/CustomClassLoader.java:
```diff
@@ -0,0 +1,27 @@
+    public Class<?> loadClass(String name) throws ClassNotFoundException {
+        if (name.startsWith(TEST_PACKAGE) || name.startsWith(MANUAL_PACKAGE)) {
+            try {
+                InputStream is = getParent().getResourceAsStream(
```
1. Please, use IOUtils.toByteArray(InputStream) here. 
2. Please also note that any input stream must be closed in finally block or by using try-with-resources.
3. Instead of empty string in ClassNotFoundException, use name of class that was not found.
4. I wonder why you use custom class loader, can you provide explanation for this decision? Add your answer to javadoc of the class.

AfanasievYuriy replied:
fixed. (additional: should we add java doc comments to all method/classes? Or only in places that you've mentioned?)

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+            List<TestClassDTO> objectList = new ArrayList<>();
+            Properties caseProperties = new Properties();
+
+            List<File> insideDirs = (List<File>) FileUtils.listFilesAndDirs(
```
It is bad practice to downcast result of external library, especially when library version is not fixed, but set to LATEST.

AfanasievYuriy replied:
fixed.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+                        TrueFileFilter.INSTANCE
+                );
+                for (File classFile : filesList) {
+                    String[] s = insideDir.getPath().split("[\\\\/]");
```
Please, add commentaries to any user regular expressions. Commentaries should provide examples of input data and result of regular expression execution.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+    private Class testClass;
+    private Class manualClass;
+
+    private static List<File> filesList = new ArrayList<>();
```
this variable can be moved into method data()

AfanasievYuriy replied:
fixed.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+                    String[] s = insideDir.getPath().split("[\\\\/]");
+                    TestClassDTO testClassDTO = new TestClassDTO(
+                            classFile.getName().split("\\.")[0],
+                            s[s.length - 2] + "/" + s[s.length - 1] + "/",
```
Using magic numbers in code is generally bad idea, please consider using apache commons FilenameUtils for common tasts like extracting file extension, or filename without extension.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+        return pog;
+    }
+
+    public MainIntegrationTest(List<TestClassDTO> list) throws MalformedURLException {
```
MalformedURLException is never thrown

AfanasievYuriy replied:
fixed.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+    }
+
+    private static List<File> getMainDirsNames() {
+        List<File> caseDirs = (List<File>) FileUtils.listFilesAndDirs(
```
Please, consider using java 7 Path api for your purposes. I would suggest to use method Files#find(Path start, int maxDepth, BiPredicate<Path, BasicFileAttributes> matcher, FileVisitOption... options). It produces cleaner code and allows to provide search parameters as lambda expressions.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+                new IOFileFilter() {
+                    @Override
+                    public boolean accept(File file) {
+                        return Pattern.compile("manual([\\\\/])[a-zA-Z]+$").matcher(file.getPath())
```
Same thing for regex: provide documentation. Also it would be good thing to move compilation of pattern out of iteration block.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+    }
+
+    private void runTestCase() throws Exception {
+        compileTestClasses();
```
Please, modify all methods to accept parameters and return values. Do not be afraid to create new model classes to store results of methods execution. To make code readable, avoid usage of testclass fields without reason. Current procedural implementation makes code unreadable, even you will not be able to support this class without strong mind effort.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+    }
+
+    private void testNameOfClass() throws Exception {
+        String[] testClassName = testClass.getName().split("\\.");
```
FileNameUtils from apache commons will help you to avoid magic numbers and make code more readable.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+    private void testAnnotationOfClass() {
+        List<Annotation> testClassAnnotations = Arrays.asList(testClass.getDeclaredAnnotations());
+        List<Annotation> manualClassAnnotations = Arrays.asList(manualClass.getDeclaredAnnotations());
+        assertEquals("Different class annotations", manualClassAnnotations,
```
I believe assertArrayEquals can be used here to avoid new ArrayList creation.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,283 @@
+
+        while (testIt.hasNext()) {
+            Field testCurrentField = testIt.next();
+            Field manualCurrentField = manualIt.next();
```
I wonder if class generator may generate class with unpredictable order of fields. If it may, then test case should be ready for that behaviour. Add javadoc comment to clarify this case.

hdghg reviewed (changes requested):
Please, dont forget to use findbugs plugin.

AfanasievYuriy committed:
Edited regexp (for universal OS file path separator). now working on win and unix

AfanasievYuriy committed:
1) edited pom (no more junit in pom.xml)
2)CustomClassLoader slightly reworked + added javadoc for class
3)Fixed unnecessary downcasting in data() method

AfanasievYuriy committed:
fixed class uploading problem (was an accident error) in CustomClassLoader

stelmakhmaksim committed:
Solve the problem with order of fields, using sorting types + names of fields.

AfanasievYuriy committed:
1) javadoc for CustomClassLoader class
2) added TestClassesDTO class for storing information about current testing and manual classes during test
3) added CompilationResultDTO for storing information about results of compilation
4) using Path API for searching manual .java files now
5) refactoring: method return values (when it's possible); MainIntegrationTest class has only 2 fields now
6) javadoc for every MainIntegrationClass method
7) solved problem with fields order of testing classes (test case)
8) compilation logic now is encapsulated into TestClassesCompiler class

AfanasievYuriy committed:
Merge branch 'EPMLSTRTAI-354' into 354-fix

# Conflicts:
#	src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java

AfanasievYuriy committed:
merge fix

Walrus94 committed:
resolved problem with different separators on windows/unix

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/CustomClassLoader.java:
```diff
@@ -0,0 +1,36 @@
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
```
Much appreciate javadoc, makes things clear

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/DTO/TestClassesDTO.java:
```diff
@@ -0,0 +1,113 @@
+ * DTO for pair of classes: manual and test (generated by POG)
+ * that are comparing to each other
+ */
+public class TestClassesDTO {
```
According to https://martinfowler.com/eaaCatalog/dataTransferObject.html DTO is pattern used to transfer data between subsystems of application (usually over network). I see in code that class only used to hold values, that is why it is not DTO.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,401 @@
+            objectList = new ArrayList<>();
+            caseProperties = new Properties();
+
+            insideCaseDirs = Files.find(
```
By trying to reusing one regex for each case you are actually making code harder to read.
This code fragment can be rewritten using following code:

    Files.list(caseDir)
                .filter(s -> Files.isDirectory(s))
                .collect(Collectors.toList());

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,401 @@
+            RESOURCE_DIR + MANUAL_DIR,
+            MAX_DEPTH_FOR_DIRS
+        );
+        List<Path> insideCaseDirs;
```
These properties can be declared at the moment of assignment

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,401 @@
+                    .matches(path) && !path.equals(caseDir)
+            ).collect(Collectors.toList());
+
+            casePropertiesList = Files.find(
```
Same problem here. Much better (and safe) approach:
            Optional<Path> propertiesFilePath = Files.list(caseDir)
                    .filter(p -> Files.isRegularFile(p))
                    .filter(p -> p.toString().endsWith(".properties"))
                    .findAny();
            if (propertiesFilePath.isPresent()) {
                caseProperties.load(Files.newInputStream(propertiesFilePath.get()));
            }

hdghg replied:
Forgot to add: Input stream should be closed

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,401 @@
+        currentClassesDTO.setTestClass(compilationResult.getExpectedClass());
+        currentClassesDTO.setManualClass(compilationResult.getManualClass());
+
+        testNameOfClass();
```
Procedure testNameOfClass here still relies on fields rather then input parameters. Make it receive parameters instead of reading test data from field.
In general, unit tests and integrational tests do not follow OOP principles, that's why it is bad idea to read/write data from fields of test case. Data should be passed from function to function instead.

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,401 @@
+     * will pass assertEquals
+     */
+    private void testNameOfClass() throws Exception {
+        String[] testClassName = currentClassesDTO
```
assertEquals("File names do not match", FilenameUtils.getName(expectedClassName)
        FilenameUtils.getName(actualClassName));

Note that expected value goes first in assertion

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,401 @@
+     * @return sorted list
+     */
+    private static List<TestClassesDTO> sortClassesList(List<TestClassesDTO> classesList) {
+        classesList.sort(
```
Comparator makes code harder to read, following snippet some easier:
    
        Map<Boolean, List<TestClassesDTO>> grouped = classesList.stream()
                .collect(Collectors.groupingBy(t -> t.getManualClassFilePath().contains("page")));
        ArrayList<TestClassesDTO> result = new ArrayList<>(grouped.get(true));
        result.addAll(grouped.get(false));
        return result;

hdghg reviewed (changes requested):
Was glad to see javadoc

AfanasievYuriy committed:
all test methods have arguments now. data pushes from method to method

IvanovIurii commented on src/test/java/com/epam/page/object/generator/integrationalTests/CustomClassLoader.java:
```diff
@@ -0,0 +1,36 @@
+                byte[] buf = IOUtils.toByteArray(is);
+                return defineClass(name, buf, 0, buf.length);
+            } catch (IOException e) {
+                throw new ClassNotFoundException(name + " class wasn't found", e);
```
Do not we need to log it?

hdghg replied:
No if it caught outside.

AfanasievYuriy committed:
EPMLSTRTAI-354: refactoring: renamed classes (package dto -> package data); search methods was simplified;

AfanasievYuriy committed:
EPMLSTRTAI-354: IS is closing now

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,368 @@
+                .filter(path -> path.toString().endsWith(".properties"))
+                .findAny();
+            if (testCasePropertyPath.isPresent()) {
+                InputStream inputStream = Files.newInputStream(testCasePropertyPath.get());
```
try-with-resources is convenient way to do this

hdghg commented on src/test/java/com/epam/page/object/generator/integrationalTests/MainIntegrationTest.java:
```diff
@@ -0,0 +1,368 @@
+    private void testModifiersOfClass(TestClassesData currentClassesDTO) {
+        int testMods = currentClassesDTO.getTestClass().getModifiers();
+        int manualMods = currentClassesDTO.getManualClass().getModifiers();
+        assertEquals("Different class modifiers", testMods, manualMods);
```
Fix all expected and actual values placement.

AfanasievYuriy committed:
EPMLSTRTAI-354: IS is closing now

AfanasievYuriy committed:
EPMLSTRTAI-354: MainTest class has been deleted (no more need, when there are MainUntegrationTest class)

hdghg committed:
EPMLSTRTAI-354: Added maven-failsafe-plugin for integration tests

AfanasievYuriy committed:
Merge remote-tracking branch 'origin/master' into EPMLSTRTAI-354

# Conflicts:
#	src/test/java/com/epam/page/object/generator/integrationalTests/MainTest.java
